docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+- Feature Name: Raft snapshots
+- Status: draft
+- Start Date: 2022-01-31
+- Authors: Denis Rystsov
+- Issue: TBD
+
+# Executive Summary
+
+Introduce Raft snapshots not bound to log truncation to solve compaction related problems with idempotency, transactions, offset translator and pave a way to custom state machines.
+
+## What is being proposed
+
+Make the snapshot before the compaction and ship it to all (majority) nodes in a cluster so that each node sees the same aggregate of the log. Currently when Redpanda does compaction it introduces holes to the log and when a follower replay the log to build an aggregate it end up with different view.
```

Comment:
Can you explain why these holes were a solution in the first place? Is it because each batch can have a different compaction timeout interval?

Reply:
I'd say not batch but a node. Each node does a compaction of its own so there maybe difference between the logs on different nodes but even if the compaction is coordinated it doesn't protect from the cases when we need to add a new node which never saw the original non compacted data.